addons/test_mail/tests/test_mail_followers_unsubscribe.py

```diff
@@ -48,3 +54,156 @@ def test_inbox_notification_follower(self):
+                    else:
+                        raise NotImplementedError(f'landing_url_check_mode {landing_url_check_mode}')
+                finally:
+                    record._message_subscribe(partner_ids=partner.ids)
+
+    @users('employee')
+    @mute_logger('odoo.addons.base.models', 'odoo.addons.mail.controllers.mail', 'odoo.http', 'odoo.models')
+    def test_unfollow_internal_user(self):
+        """Internal user must receive an unfollow URL, that works, cannot be tampered and redirects to the correct page.
+
+        Note: When connected, the user must be redirected to the record otherwise a simple front end page.
+        """
+        test_record = self.test_record_admin
+        test_partner = self.partner_employee
+
+        test_record._message_subscribe(partner_ids=test_partner.ids)
+        try:
+            with self.subTest('Internal user receives unfollow URL'):
+                unfollow_url = self._extract_unfollow_url(self._message_post(test_record))
+                self.assertTrue(unfollow_url)
+            self._test_unfollow_url(test_record, unfollow_url, test_partner, 'front_end_message')
+            self._test_tampered_unfollow_url(test_record, unfollow_url, test_partner)
+
+            self.authenticate(self.user_employee.login, self.user_employee.login)
+            self._test_unfollow_url(test_record, unfollow_url, test_partner, 'record')
+
+            test_record.message_unsubscribe(partner_ids=test_partner.ids)
+            with self.subTest('Internal user simple notification (without unfollow URL)'):
+                unfollow_url = self._extract_unfollow_url(self._notify_thread_by_email(test_record, test_partner))
+                self.assertFalse(unfollow_url)
+        finally:
+            test_record.message_unsubscribe(partner_ids=test_partner.ids)
+
+    @mute_logger('odoo.models')
+    def test_unfollow_partner_with_no_user(self):
+        """External partner must not receive an unfollow URL."""
+        test_record = self.test_record_admin
+        test_partner = self.partner_without_user
+
+        test_record._message_subscribe(partner_ids=test_partner.ids)
+        try:
+            with self.subTest('External partner must not receive an unfollow URL'):
+                unfollow_url = self._extract_unfollow_url(self._message_post(test_record))
+                self.assertFalse(unfollow_url)
+        finally:
+            test_record.message_unsubscribe(partner_ids=self.partner_without_user.ids)
+
+    @mute_logger('odoo.addons.mail.controllers.mail', 'odoo.models', 'odoo.http')
+    def test_unfollow_partner_with_no_user_on_record_unfollow_enabled(self):
+        """External partner must receive an unfollow URL for message related to record with unfollow enabled."""
+        test_record = self.test_record_unfollow_admin
+        test_partner = self.partner_without_user
+
+        test_record._message_subscribe(partner_ids=test_partner.ids)
+        try:
```

Comment:
why this try / finally ?

Reply:
The idea is to restore the initial state for other test so that a failure of a test doesn’t impact another test.

Reply:
but still, the TransactionCase will rollback the transaction at the end of the method right ? so you shouldn't need to call `_message_subscribe`

(it's useful if you loop and test multiple things in the same method, but here the transaction should be rollbacked isn't it ?)

Reply:
Yes you’re right. I have removed it from all the tests but kept in the helper function as it execute multiple tests in the same test (same transaction). 
My confusion cames from the fact that in such test, the setup is only executed once for all test,  TransactionCase is only inherited indirectly and I saw such state restoration in other test. So thank you for the explanation.